lang/cs/Org.Apache.REEF.Network/NetworkService/WritableNsMessage.cs

```diff
@@ -0,0 +1,184 @@
+ * under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Hadoop.Avro;
+using Org.Apache.REEF.Network.Group.Driver.Impl;
+using Org.Apache.REEF.Wake;
+using Org.Apache.REEF.Wake.Remote;
+
+namespace Org.Apache.REEF.Network.NetworkService
+{
+    /// <summary>
+    /// Writable Message sent between NetworkServices.</summary>
+    /// <typeparam name="T">The type of data being sent. It is assumed to be Writable</typeparam>
+    [Obsolete("Need to remove Iwritable and use IstreamingCodec. Please see Jira REEF-295 ", false)]
+    public class WritableNsMessage<T> : IWritable where T : IWritable
+    {
+        private string _sourceIdString;
+        private string _destIdString;
+
+        /// <summary>
+        /// Constructor to allow instantiation by reflection
+        /// </summary>
+        public WritableNsMessage()
+        {
+        }
+        
+        /// <summary>
+        /// Create a new Writable NsMessage with no data.
+        /// </summary>
+        /// <param name="sourceId">The identifier of the sender</param>
+        /// <param name="destId">The identifier of the receiver</param>
+        public WritableNsMessage(IIdentifier sourceId, IIdentifier destId)
+        {
+            SourceId = sourceId;
+            DestId = destId;
+            Data = new List<T>();
+        }
+
+        /// <summary>
+        /// Create a new Writable NsMessage with data.
+        /// </summary>
+        /// <param name="sourceId">The identifier of the sender</param>
+        /// <param name="destId">The identifier of the receiver</param>
+        /// <param name="message">The message to send</param>
+        public WritableNsMessage(IIdentifier sourceId, IIdentifier destId, T message)
+        {
+            SourceId = sourceId;
+            DestId = destId;
+            Data = new List<T> {message};
+        }
+
+        /// <summary>
+        /// The identifier of the sender of the message.
+        /// </summary>
+        internal IIdentifier SourceId { get; private set; }
+
+        /// <summary>
+        /// The identifier of the receiver of the message.
+        /// </summary>
+        internal IIdentifier DestId { get; private set; }
+
+        /// <summary>
+        /// A list of data being sent in the message.
+        /// </summary>
+        public List<T> Data { get; set; }
+
+        /// <summary>
+        /// Read the class fields.
+        /// </summary>
+        /// <param name="reader">The reader from which to read </param>
+        public void Read(IDataReader reader)
+        {
+            _sourceIdString = reader.ReadString();
+            _destIdString = reader.ReadString();
+            int messageCount = reader.ReadInt32();
+
+            Data = new List<T>();
+
+            for (int index = 0; index < messageCount; index++)
+            {
```

Comment:
Inside of the loop, there are two indexed accesses to the list which could be avoided.